### ISSUE TITLE
YM2610 - Clicky/poppy sound in Ganryu

### ISSUE THREAD

**Issue by carmiker** [aaronsgiles/ymfm#58]
I noticed in Ganryu (Neo Geo), there is a noticeable click/pop sound when slashing, and in a few other places where it is a little less noticeable. This does not happen on real hardware (I own the game and 3 MVS boards and can confirm). It also does not happen on FinalBurn Neo, which uses code derived from the old MAME YM2610.

My recording using my Neo Geo emulator ([Geolith](https://gitlab.com/jgemu/geolith)):
https://files.catbox.moe/w29l21.ogg
This uses an admittedly very modified YMFM, but it sounds the same in the most recent MAME revision I am able to build on my musl-libc system (0.256).

Example on real hardware:
https://youtu.be/-DCWWSm5N0c?t=167

I hope this is enough to begin a diagnosis.

Thanks!

**Comment by happppp** [comment]
The sample is from adpcm_a, channel mask 0x08. The click is caused by ADPCM accumulator overflow.

![Image](https://github.com/user-attachments/assets/4fea830f-0714-44b8-97b5-aeb12699f647)

For your NeoGeo emulator, you could circumvent it like this:
replace ymfm_adpcm.cpp line 207 where it claims accumulator wraps:
m_accumulator = clamp(m_accumulator + delta, -2048, 2047);

I can't hear from that video whether or not this sample has the click.
To prove that it doesn't happen on PCB, I think a phone camera microphone is not sufficient, with bonus background noise and talking.

**Comment by carmiker** [comment]
@happppp, Thanks for pointing out the source of the issue.

You are correct, the youtube video is not sufficient. I am not sure why I didn't notice it when I first tested on real hardware, maybe I didn't listen closely enough or have the volume high enough.

There is a similar issue on the newly released Shinobi port, also the same between ymfm and real hardware: https://h0ffman.itch.io/shinobi-neogeo

I think this issue can be closed now.

**Comment by happppp** [comment]
So, you're saying the click in Ganryu's slash sample does happen on PCB? Not by assumption from a similar sounding bug in Shinobi, but you've heard/measured it? (you said so, but then edited your post)

**Comment by carmiker** [comment]
Yes, the click does happen in Ganryu's slash sample and others on real hardware. The code change you suggested "fixes" all the clicks and pops in both Ganryu and Shinobi, but Shinobi exhibits the problem on real hardware as well. I believe it's the same root cause in both cases.

HOWEVER... the clicks in the Ganryu character select screen do NOT happen on real hardware. So, I am reopening this as it's not fully resolved yet. I don't have the means to take a direct recording from the hardware at the moment.

**Comment by carmiker** [comment]
I managed to get the cable required to input the raw audio into one of my computers. The wires are directly connected to the stereo output pins on an MV1C in this case.

Ganryu: https://files.catbox.moe/wd0xv7.ogg
- There are pops in the character select screen background music (12-24 seconds).
- There are clicks when slashing the sword (30-34 seconds).
- There are clicks when sliding (35-40 seconds).

Shinobi: https://files.catbox.moe/nej6cg.ogg
- There are scratches when the "MISSION 1" voice sample plays (6-8 seconds).

Conclusion: There is nothing wrong with ymfm in this regard. Listening on low quality integrated Sony PVM speakers is a losing endeavour. Clamping, as suggested in the above comment, will eliminate the pops exhibited on real hardware if someone wanted to implement a QOL hack.
